[typing] define additional methods & properties on jax.Array

jakevdp opened the pull request:
These are the methods that are only valid for actual materialized arrays (i.e. not Tracers) In order to simplify the experience for users, we want to maintain only a single jax.Array type, so we define all methods here and raise explicit errors on Tracer instances.

Part of #12049

yashk2810 commented on jax/_src/basearray.pyi:
```diff
@@ -186,3 +186,24 @@ class Array(abc.ABC):
+  @property
+  def device_buffer(self) -> Array: ...
+  @property
+  def device_buffers(self) -> Sequence[Array]: ...
```
Do we want to expose these? These methods are deprecated `device_buffer` and `device_buffers`

jakevdp replied:
Makes sense, yeah we should probably leave these out. Are there any other deprecated methods/properties in this list?

yashk2810 replied:
Rest of them are good!

jakevdp committed:
[typing] define additional methods & properties on jax.Array

These are the methods that are only valid for actual materialized arrays (i.e. not Tracers)
In order to simplify the experience for users, we want to maintain only a single jax.Array
type, so we define all methods here and raise explicit errors on Tracer instances.
